### PR TITLE
Add nanogpt and training function

### PR DESCRIPTION
**Pull request by sophgen**
Adapted NanoGPT and make modifications in order to work with TextDataLoader classes

### CODE REVIEW

File: experiments/reconstruct_shakespeare.py [hunk 1]
```diff
@@ -35,6 +39,11 @@
     parser.add_argument("--nhead", type=int, default=8, help="Number of attention heads")
     parser.add_argument("--num_decoder_layers", type=int, default=6, help="Number of decoder layers")
     parser.add_argument("--dim_feedforward", type=int, default=2048, help="Dimension of the feedforward network")
+    parser.add_argument("--nembd", type=int, default=768, help="Dimension of the embedding layer for NanoGPT")
+    parser.add_argument("--nlayer", type=int, default=12, help="Number of layers for NanoGPT")
```
Comment (nateburley) [inline experiments/reconstruct_shakespeare.py:43]: Let's get rid of num_encoder_layers and num_decoder_layers and just substitute this in to make it consistent

File: experiments/reconstruct_shakespeare.py [hunk 2]
```diff
@@ -35,6 +39,11 @@
...
+    parser.add_argument("--block_size", type=int, default=1024, help="Block size for NanoGPT")
+    parser.add_argument("--dropout", type=float, default=0.0, help="Drop out")
```
Comment (nateburley) [inline experiments/reconstruct_shakespeare.py:45]: Dropout is in already, might need to merge with main

File: utils/training.py [hunk 3]
```diff
@@ -133,6 +136,125 @@ def train_shakespeare_transformer(model, train_loader, eval_loader, optimizer, n
     plt.show()
 
 
+def train_nanogpt(model, device, train_data_loader, val_data_loader, max_iters, batch_size):
```
Comment (nateburley) [inline utils/training.py:139]: Where does the chunking and context window splitting happen in the training function?

File: utils/training.py [hunk 4]
```diff
@@ -133,6 +136,125 @@ def train_shakespeare_transformer(model, train_loader, eval_loader, optimizer, n
...
+
+    # helps estimate an arbitrarily accurate loss over either split using many batches
+    @torch.no_grad()
+    def estimate_loss():
+        out = {}
+        model.eval()
+        for split in ['train', 'val']:
+            losses = torch.zeros(eval_iters)
+            for k in range(eval_iters):
+                X, Y = get_batch(split)
```
Comment (nateburley) [inline utils/training.py:179]: I think we need to add the logic here to split the chunk into a piece of context_size, right?

**Review by nateburley — CHANGES_REQUESTED** [review]
Looks good overall! The only possible change I see is that we might need to tweak the training loop though to split the chunks into sequences of length context_window, and then iterate through the chunk

**Comment by sophgen** [comment]
Thanks so much @nateburley  for the comments, I will close this one and create a new PR from latest main. Thanks!
